### ISSUE TITLE
Can't get Google Test unit tests to show up

### ISSUE THREAD

**Issue by mdeforge** [fredericbonnet/cmake-test-explorer#30]
I can't get my Google Test unit tests to show up in the Testing extension window. If I use C++ TestMate I can, but not CMake Test Explorer. I guess what are the differences between the two? Is this expected? If so, maybe I'm misunderstanding the purpose of CMake Test Explorer vs C++ TestMate. If it's not, maybe it's my usage of gtest_discover_tests() instead of add_test()? I know the former calls the later but maybe there's an incompatibility there. Any advice would be appreciated.

**Comment by fredericbonnet** [comment]
Hi! A few questions:

- What is the version number of CMake/CTest?
- What is the version number of the extension? I updated the extension 2 days ago
- Does the test explorer panel display an error or is it just empty? 
- Is there any error message on any of the log/output panels?

In any case this extension should work the same regardless of the test declaration method, as it simply calls CTest directly. The only difference between C++ TestMate and this extension is the supported frameworks (GTest/Catch2 etc. for the former, CTest for the later), apart from that they rely on the same Test Explorer UI.

**Comment by mdeforge** [comment]
Hey! I'm using CMake 3.18.4, and CMake Test Explorer v0.14.0. Yeah the panel is just empty. No error displayed in the testing panel or in the log/output panels from the build.

Other potential factors -- I just have the 2019 Build Tools installed, I don't have Visual Studio itself installed. Everything else seems to be happy with this though (I am compiling and debugging fine).

In my project I'm including CTest and then adding the tests subdirectory. In the tests project I include GoogleTest, obtain GoogleTest via FetchContent, set gtest_force_shared_crt to on, link with gtest_main. and then call gtest_discover_tests().

**Comment by fredericbonnet** [comment]
Everything looks fine to me. Can you try running the following command from your build dir:

```sh
ctest --show-only=json-v1
```

This is the command that the extension runs to get the test list.

You also have to make sure that the `cmakeExplorer.buildDir` settings is set to the right location (unless you use the CMake Tools extension as well, in which case it should happen automatically).

Also, is it a fresh installation of the extension or has it just stopped working with the new update?

**Comment by mdeforge** [comment]
Here's the result:
```
{
  "backtraceGraph" :
  {
    "commands" : [],
    "files" : [],
    "nodes" : []
  },
  "kind" : "ctestInfo",
  "tests" : [],
  "version" :
  {
    "major" : 1,
    "minor" : 0
  }
}
```
`cmakeExplorer.buildDir` was set to ${buildDirectory} which was empty, so I changed that to "${CMAKE_PROJECT_DIR}" but that I got an error in the CMake Test Explorer saying it couldn't find CMakeCache.txt. That was weird, so I checked the path and it was still wrong so I just put in "build". That fixed the error, but I still wasn't seeing anything.

On a hunch, I moved my test from inside the nested project folders to the top level. So, instead of Foo Project\Bar Library\Bar Library Tests, I put it at Foo Project\Bar Library Tests. I then saw it appear in the CMake Test Explorer panel! I ran that command again and was surprised to see it still printed out the same thing as above. Was kind of expecting a change after it was working.

So does this extension not drill down recursively into the projects then? Do I need to have my tests at the top level?

**Comment by fredericbonnet** [comment]
The extension tries to locate the `CMakeCache.txt` file at the root of the `buildDir` directory. `${buildDirectory}` is a special variable that's mapped to a CMake Tools extension setting (see https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-settings.md). If this extension is not installed then it defaults to empty, and `buildDir` will point to the workspace root.

When `buildDir` has default or empty value the extension fails silently if the `CMakeCache.txt` doesn't exist in the workspace root. This is to prevent unwanted errors on projects that don't use CMake if the extension is enabled globally. However if you change the `buildDir` value then the extension will expect the `CMakeCache.txt` file to exist, and will display an error if it doesn't. Hence the difference between `${buildDirectory}` and  `${CMAKE_PROJECT_DIR}`.

Another point is that you have to prefix environment variables with `env:` in variable substitutions, so here you need `${env:CMAKE_PROJECT_DIR}` if you want to use that variable.

> On a hunch, I moved my test from inside the nested project folders to the top level. So, instead of Foo Project\Bar Library\Bar Library Tests, I put it at Foo Project\Bar Library Tests. I then saw it appear in the CMake Test Explorer panel! I ran that command again and was surprised to see it still printed out the same thing as above. Was kind of expecting a change after it was working.

I'm a bit surprised by that. The extension simply runs the `ctest --show-only=json-v1` command from `buildDir`, so you should get identical results in both cases. Maybe specifying the config would help, i.e. `--build-config=debug`?

> So does this extension not drill down recursively into the projects then? Do I need to have my tests at the top level?

No, it only runs from `buildDir`, however if your project structure uses subdirs and they are properly configured in the `CMakeLists.txt` files then they should show up properly with `ctest`. The extension does nothing clever with CMake files and rely on `ctest` alone to avoid any surprising behavior.

**Comment by mdeforge** [comment]
While writing the following, somehow things started working. I didn't change any of the default settings, and `ctest --show-only=json-v1` still shows the same empty graph. I think I'm expecting to see my example test in there? Adding `--build-config=debug` made no difference. Weird.

At any rate, I still think what I was writing is worth the read. I suspect that either CMake Tools was not grabbing fresh data, VS Code itself was not giving you fresh data, or CMake was not refreshing it's cache to provide the latest data. **I was running "CMake: Delete Cache and Reconfigure" a lot there before it magically started working.**

---

I have the CMake Tools extension installed (v1.7.3) and Cmake: Build Directory is set to `${workspaceFolder}/build`. That location is indeed my build folder and it does contain CMakeCache.txt.

I created a task that echo's the current value `${workspaceFolder}`

```
{
    "label": "Workspace Echo",
    "type": "shell",
    "command": "echo ${workspaceFolder}/build"
}
```
The output is `Z:\Foo/build` as I would expect.

`${command:cmake.buildDirectory}` echos `z:/Foo/build`, as I would expect as well.

These are all default values I'm using. I'm not changing anything in settings.

---

Thanks for the help! I think I can mark this as closed now.

**Comment by fredericbonnet** [comment]
That's good to know, thanks!

**Comment by IngmarPaetzold** [comment]
Hi there, in case someone stumbles over this thread because the tests are not shown, this helped me. (I was re-arranging test directories, and suddenly the test runner was empty, even after I recovered to the last git version where it used to work. That was really annoying).
The reason was: in the main `CMakeLists.txt` file, the order of the following commands was wrong. This is how it needs to be:

    enable_testing()
    add_subdirectory(<your test dir>)

Credits: found the hint here: https://stackoverflow.com/questions/30250494/ctest-not-detecting-tests

**Comment by geofflangdale** [comment]
Annoyingly, I tried reordering a bunch of things, and had a very similar experience of things 'magically starting working again'. The last thing I changed was to move enable_testing and include(GoogleTest) to a top-level CMakeLists.txt, but I'm not confident that was really it. A simple change in the order (as per Ingmar's comment above) was **not** sufficient to fix things for me. However, I have no idea what I'm doing with CMake and even less idea of what I'm doing with CMake+Visual Studio.
